File: src/LargeFamily.cpp
#include "../include/LargeFamily.h"
#include "../include/VCFHeteroHomo.h"
#include "../include/VCFFillable.h"
#include "../include/VCFImpFamily.h"
#include "../include/VCFJunkRecord.h"
#include "../include/VCFHomoHomo.h"
#include "../include/VCFHeteroHeteroLite.h"
#include "../include/Pedigree.h"
#include "../include/KnownFamily.h"
#include "../include/Map.h"
#include "../include/option.h"
#include "../include/common.h"

using namespace std;


//////////////////// LargeFamily ////////////////////

VCFHeteroHomoRecord *LargeFamily::create_heterohomo_record(const STRVEC& v,
										const KnownFamily *family, size_t i,
										WrongType wrong_type, ParentComb pc) {
	const STRVEC&	samples = family->get_samples();
	const int	total_int_gt = pc == ParentComb::P00x01 ? 1 : 3;
	if(!family->is_mat_known()) {
		const int	pat_int_gt = Genotype::get_int_gt(v[10]);
		const int	mat_int_gt = total_int_gt - pat_int_gt;
		if(mat_int_gt < 0 || 2 < mat_int_gt)
			wrong_type = WrongType::MODIFIABLE;
		else {
			STRVEC	w(v.size());
			std::copy(v.begin(), v.end(), w.begin());
			w[9] = Genotype::int_to_gt(mat_int_gt);
			return new VCFHeteroHomoRecord(w, samples, i, wrong_type, pc);
		}
	}
	else if(!family->is_pat_known()) {
		const int	mat_int_gt = Genotype::get_int_gt(v[9]);
		const int	pat_int_gt = total_int_gt - mat_int_gt;
		if(pat_int_gt < 0 || 2 < pat_int_gt)
			wrong_type = WrongType::MODIFIABLE;
		else {
			STRVEC	w(v.size());
			std::copy(v.begin(), v.end(), w.begin());
			w[10] = Genotype::int_to_gt(pat_int_gt);
			return new VCFHeteroHomoRecord(w, samples, i, wrong_type, pc);
		}
	}
	
	return new VCFHeteroHomoRecord(v, samples, i, wrong_type, pc);
}

void LargeFamily::classify_record(size_t i, const VCFFamily *vcf,
								  const TypeDeterminer *td,
								  const KnownFamily *family,
								  vector<VCFHeteroHomoRecord *>& heho_records,
								  vector<VCFImpFamilyRecord *>& other_records) {
	ClassifyRecord	*CR = ClassifyRecord::get_instance();
	VCFFamilyRecord	*record = vcf->get_family_record(i);
	const auto	pair1 = CR->classify(record, td, family->is_one_unknown());
	const ParentComb	pc = pair1.first;
	const WrongType	wrong_type = pair1.second;
	const STRVEC&	v = record->get_v();
	const STRVEC&	samples = record->get_samples();
	if(pc == ParentComb::PNA) {		// no candidates
		other_records[i] = new VCFJunkRecord(v, samples, i, wrong_type);
	}
	else if(TypeDeterminer::is_homohomo(pc)) {
		auto	*r_ = new VCFHomoHomoRecord(v, samples, i, wrong_type, pc);
		auto	*r = r_->impute();
		delete r_;
		other_records[i] = r;
	}
	else if(TypeDeterminer::is_heterohomo(pc)) {
		heho_records[i] = create_heterohomo_record(v, family,
													i, wrong_type, pc);
	}
	else {
		other_records[i] = new VCFHeteroHeteroLiteRecord(v, samples, i,
															wrong_type, pc);
	}
}

void LargeFamily::classify_records_in_thread(void *config) {
	auto	*c = (ConfigThreadClassify *)config;
	const size_t	n = c->vcf->size();
	for(size_t i = c->first; i < n; i += c->num_threads) {
		classify_record(i, c->vcf, c->td, c->family,
							c->heho_records, c->other_records);
	}
}

pair<vector<VCFHeteroHomoRecord *>, vector<VCFImpFamilyRecord *>>
LargeFamily::classify_records(const VCFFamily *vcf,
							  const KnownFamily *family, const Option *option) {
	const size_t	N = vcf->size();
	vector<VCFHeteroHomoRecord *>	heho_records(N, NULL);
	vector<VCFImpFamilyRecord *>	other_records(N, NULL);
	ClassifyRecord	*CR = ClassifyRecord::get_instance();
	const auto	*td = CR->get_TypeDeterminer(vcf->num_samples()-2,
														option->ratio);
	const int	T = vcf->size() < 100 ? 1 : option->num_threads;
	
	vector<ConfigThreadClassify *>	configs(T);
	for(int i = 0; i < T; ++i)
		configs[i] = new ConfigThreadClassify(i, T, vcf, td, family,
											  heho_records, other_records);
	
#ifndef DEBUG
	vector<pthread_t>	threads_t(T);
	for(int i = 0; i < T; ++i)
		pthread_create(&threads_t[i], NULL,
			(void *(*)(void *))&classify_records_in_thread, (void *)configs[i]);
	
	for(int i = 0; i < T; ++i)
		pthread_join(threads_t[i], NULL);
#else
	for(int i = 0; i < T; ++i)
		classify_records_in_thread(configs[i]);
#endif
	
	Common::delete_all(configs);
	
	return std::make_pair(heho_records, other_records);
}

void LargeFamily::create_vcf_in_thread(void *config) {
	auto	*c = (ConfigThreadCreate *)config;
	const VCFSmall	*orig_vcf = c->orig_vcf;
	for(size_t i = c->first; i < c->families.size(); i += c->num_threads) {
		const KnownFamily	*family = c->families[i];
		c->results[i] = VCFFamily::create(orig_vcf, family->get_samples());
	}
}

vector<VCFFamily *> LargeFamily::create_family_vcfs(
								const VCFSmall *orig_vcf,
								const vector<const KnownFamily *>& families,
								int num_threads) {
	const size_t	N = families.size();
	vector<VCFFamily *>	vcfs_family(N, NULL);
	const int	T = std::min((int)N, num_threads);
	vector<ConfigThreadCreate *>	configs(T);
	for(int i = 0; i < T; ++i)
		configs[i] = new ConfigThreadCreate(i, T, orig_vcf,
											families, vcfs_family);
	
#ifndef DEBUG
	vector<pthread_t>	threads_t(T);
	for(int i = 0; i < T; ++i)
		pthread_create(&threads_t[i], NULL,
			(void *(*)(void *))&create_vcf_in_thread, (void *)configs[i]);
	
	for(int i = 0; i < T; ++i)
		pthread_join(threads_t[i], NULL);
#else
	for(int i = 0; i < T; ++i)
		create_vcf_in_thread(configs[i]);
#endif
	
	Common::delete_all(configs);
	
	return vcfs_family;
}

vector<pair<string, vector<pair<size_t, size_t>>>>
LargeFamily::collect_same_parent_families(
								const vector<const KnownFamily *>& families) {
	// { parent: [(family index, parent index)] }
	map<string, vector<pair<size_t, size_t>>>	dic;
	for(size_t i = 0; i < families.size(); ++i) {
		dic[families[i]->get_mat()].push_back(make_pair(i, 0));
		dic[families[i]->get_pat()].push_back(make_pair(i, 1));
	}
	
	vector<pair<string, vector<pair<size_t, size_t>>>>	v;
	for(auto p = dic.begin(); p != dic.end(); ++p) {
		if(p->second.size() > 1)
			v.push_back(*p);
	}
	return v;
}

vector<vector<VCFImpFamilyRecord *>> LargeFamily::collect_same_position_records(
				const vector<tuple<vector<VCFHeteroHomoRecord *>,
								   vector<VCFImpFamilyRecord *>, int>>& rs) {
	const size_t	vcf_size = get<0>(rs[0]).size();
	vector<vector<VCFImpFamilyRecord *>>	recordss(vcf_size);
	for(auto p = rs.begin(); p != rs.end(); ++p) {
		const auto&	rs1 = get<0>(*p);
		for(auto q = rs1.begin(); q != rs1.end(); ++q) {
			auto	*r1 = *q;
			if(r1 != NULL && r1->is_right())
				recordss[r1->get_index()].push_back(r1);
		}
		const auto&	rs2 = get<1>(*p);
		for(auto q = rs2.begin(); q != rs2.end(); ++q) {
			auto	*r2 = *q;
			if(r2 != NULL && r2->is_homohomo())
				recordss[r2->get_index()].push_back(r2);
		}
	}
	return recordss;
}

void LargeFamily::modify_00x11_each(
				const vector<tuple<vector<VCFHeteroHomoRecord *>,
								   vector<VCFImpFamilyRecord *>, int>>& rs) {
	auto	records = collect_same_position_records(rs);
	
	for(auto p = records.begin(); p != records.end(); ++p) {
		if(p->size() >= 2)
			VCFImpFamilyRecord::modify_00x11(*p);
	}
}

// When a parent of record with 0/0 x 1/1 parents is one of another family,
// correct if the genotype is different from other Homo x Homo or Hetero x Homo
void LargeFamily::modify_00x11(
					vector<vector<VCFHeteroHomoRecord *>>& heho_recordss,
				 	vector<vector<VCFImpFamilyRecord *>>& other_recordss,
				 	const vector<const KnownFamily *>& families) {
	auto	fams = collect_same_parent_families(families);
	for(auto p = fams.begin(); p != fams.end(); ++p) {
		auto	v = p->second;
		vector<tuple<vector<VCFHeteroHomoRecord *>,
					 vector<VCFImpFamilyRecord *>, int>>	rs;
		for(auto q = v.begin(); q != v.end(); ++q) {
			const size_t	fam_index = q->first;
			const size_t	p_index = q->second;
			rs.push_back(make_tuple(heho_recordss[fam_index],
									other_recordss[fam_index], p_index));
		}
		modify_00x11_each(rs);
	}
}

// divide records into hetero x homo and other types
pair<vector<vector<VCFHeteroHomoRecord *>>,
	 vector<vector<VCFImpFamilyRecord *>>>
LargeFamily::divide_vcf_into_record_types(
								const vector<VCFFamily *>& family_vcfs,
								const vector<const KnownFamily *>& families,
								const Option *option) {
	vector<vector<VCFHeteroHomoRecord *>>	heho_recordss;
	vector<vector<VCFImpFamilyRecord *>>	other_recordss;
	for(size_t i = 0; i < family_vcfs.size(); ++i) {
		auto	*vcf = family_vcfs[i];
		auto	*family = families[i];
		auto	pair1 = classify_records(vcf, family, option);
		auto	heho_records = pair1.first;
		auto	other_records = pair1.second;
		heho_recordss.push_back(heho_records);
		other_recordss.push_back(other_records);
	}
	return make_pair(heho_recordss, other_recordss);
}

void LargeFamily::fill_in_thread(void *config) {
	const auto	*c = (ConfigThreadFill *)config;
	const size_t	n = c->vcfss_heho.size();
	for(size_t i = c->first; i < n; i += c->num_threads) {
		const auto&	vcfs = c->vcfss_heho[i];
		const auto&	records = c->other_recordss[i];
		c->results[i] = VCFFillable::fill(vcfs, records,
											c->num_threads_in_family);
	}
}

vector<VCFFillable *> LargeFamily::fill_vcf(
					const map<string, vector<VCFHeteroHomo *>>& dic_vcfs,
					const vector<vector<VCFImpFamilyRecord *>>& other_recordss,
					const vector<const KnownFamily *>& families,
					int num_threads) {
	// collect vcf by family index
	const size_t	N = families.size();
	map<pair<string, string>, size_t>	family_indices;
	for(size_t i = 0; i < N; ++i)
		family_indices.insert(make_pair(families[i]->parents(), i));
	vector<vector<VCFHeteroHomo *>>	vcfss_heho(N);
	for(auto p = dic_vcfs.begin(); p != dic_vcfs.end(); ++p) {
		const auto&	vcfs = p->second;
		for(auto q = vcfs.begin(); q != vcfs.end(); ++q) {
			auto	*vcf = *q;
			const size_t	index = family_indices[vcf->parents()];
			vcfss_heho[index].push_back(vcf);
		}
	}
	
	// insert and correct other records
	vector<VCFFillable *>	vcfs_filled(N);
	const int	T = std::min(num_threads, (int)N);
	const int	T_in_family = std::max(2, (T+(int)N-1)/(int)N);
	vector<ConfigThreadFill *>	configs(T);
	for(int i = 0; i < T; ++i)
		configs[i] = new ConfigThreadFill(i, T, vcfss_heho, other_recordss,
													T_in_family, vcfs_filled);
	
#ifndef DEBUG
	vector<pthread_t>	threads_t(T);
	for(int i = 0; i < T; ++i)
		pthread_create(&threads_t[i], NULL,
					(void *(*)(void *))&fill_in_thread, (void *)configs[i]);
	
	for(int i = 0; i < T; ++i)
		pthread_join(threads_t[i], NULL);
#else
	for(int i = 0; i < T; ++i)
		fill_in_thread(configs[i]);
#endif
	
	Common::delete_all(configs);
	return vcfs_filled;
}

void LargeFamily::compress_records(vector<VCFImpFamilyRecord *>& others) {
	auto	p = others.begin();
	for(auto q = others.begin(); q != others.end(); ++q) {
		if(*q != NULL) {
			*p = *q;
			++p;
		}
	}
	others.erase(p, others.end());
}

void LargeFamily::clean_in_thread(void *config) {
	auto	*c = (ConfigThreadClean *)config;
	const int	n = (int)c->families.size();
	const int	T = c->num_threads;
	const int	T_in_family = std::max(2, (T+n-1)/n);
	for(int i = c->first; i < n; i += T) {
		const auto&	records = c->recordss[i];
		const auto&	samples = c->families[i]->get_samples();
		auto	header = c->orig_vcf->trim_header(samples);
		auto	p = VCFHeteroHomo::clean_vcfs(records, header, samples,
													c->geno_map, T_in_family);
		c->vcfss[i] = p.first;
		c->unused_recordss[i] = p.second;
	}
}

VCFSmall *LargeFamily::correct_large_family_VCFs(
									const VCFSmall *orig_vcf,
									const vector<const KnownFamily *>& families,
									const Map& geno_map, const Option *option) {
	const int	num_threads = option->num_threads;
	// create a VCF for each large family
	const auto	family_vcfs = create_family_vcfs(orig_vcf,
												families, num_threads);
	// classify records
	auto	p = divide_vcf_into_record_types(family_vcfs, families, option);
	auto&	heho_recordss = p.first;
	auto&	other_recordss = p.second;
	
	// We have to deal with it all together
	// because we coordinate between families.
	modify_00x11(heho_recordss, other_recordss, families);
	
	// impute VCFs by family
	const size_t	LN = families.size();
	vector<vector<VCFHeteroHomo *>>	vcfss(LN);
	vector<vector<VCFHeteroHomoRecord *>>	unused_recordss(LN);
	const int	T = min((int)LN, num_threads);
	vector<ConfigThreadClean *>	configs(T);
	for(int i = 0; i < T; ++i)
		configs[i] = new ConfigThreadClean(i, T, heho_recordss,
										   orig_vcf, families,
										   geno_map, vcfss, unused_recordss);
	
#ifndef DEBUG
	vector<pthread_t>	threads_t(T);
	for(int i = 0; i < T; ++i)
		pthread_create(&threads_t[i], NULL,
					(void *(*)(void *))&clean_in_thread, (void *)configs[i]);
	
	for(int i = 0; i < T; ++i)
		pthread_join(threads_t[i], NULL);
#else
	for(int i = 0; i < T; ++i)
		clean_in_thread(configs[i]);
#endif
	
	Common::delete_all(configs);
	
	for(size_t i = 0; i < families.size(); ++i) {
		auto&	others = other_recordss[i];
		const auto&	unused = unused_recordss[i];
		compress_records(others);
		others.insert(others.end(), unused.begin(), unused.end());
	}
	
	Common::delete_all(family_vcfs);
	
	// collect VCFs by parent and decide phasing
	map<string, vector<VCFHeteroHomo *>>	dic_vcfs;
	for(size_t i = 0; i < families.size(); ++i) {
		auto	vcfs = vcfss[i];
		const Family	*family = families[i];
		for(auto p = vcfs.begin(); p != vcfs.end(); ++p) {
			auto	*vcf = *p;
			if(vcf->size() == 0) {
				// Add a VCF to a dic_vcfs even if it is empty,
				// because if no VCF exists, you won't know the family
				dic_vcfs[family->get_mat()].push_back(vcf);
				dic_vcfs[family->get_pat()].push_back(vcf);
			}
			else if(vcf->is_mat_hetero())
				dic_vcfs[family->get_mat()].push_back(vcf);
			else
				dic_vcfs[family->get_pat()].push_back(vcf);
		}
	}
	
	for(auto p = dic_vcfs.begin(); p != dic_vcfs.end(); ++p) {
		auto	vcfs = p->second;
		VCFHeteroHomo::inverse_phases(vcfs);
	}
	
	auto	vcfs_filled = fill_vcf(dic_vcfs, other_recordss,
										families, option->num_threads);
	for(size_t i = 0; i < families.size(); ++i) {
		Common::delete_all(vcfss[i]);
		Common::delete_all(other_recordss[i]);
	}
	auto	vcf1 = VCFFillable::merge(vcfs_filled, orig_vcf->get_samples());
	for(auto p = vcfs_filled.begin(); p != vcfs_filled.end(); ++p) {
		delete *p;
	}
	return vcf1;
}

File: python/LargeFamily.py
from __future__ import annotations

# coding: utf-8
# LargeFamily.py

from itertools import *
from collections import defaultdict, Counter
import sys
from multiprocessing import Pool

from typing import Optional

from VCF import *
from VCFGeno import VCFGeno
from VCFFamily import VCFFamily, VCFFamilyRecord
from pedigree import PedigreeTable, Family
from VCFHeteroHomo import *
from VCFHomoHomo import VCFHomoHomoRecord
from VCFHeteroHeteroLiteRecord import VCFHeteroHeteroLiteRecord
from VCFJunkRecord import *
from VCFFillable import *
from VCFHeteroHomoPP import *
from VCFImpFamilyRecord import VCFImpFamilyRecord
from Map import *
import ClassifyRecord as CR
from KnownFamily import KnownFamily
from TypeDeterminer import TypeDeterminer, ParentComb
from Genotype import Genotype
from option import *
from common import *


#################### process ####################

def create_heterohomo_record(pos: int, geno: list[int], family: KnownFamily,
			i: int, wrong_type: str, pair: ParentComb) -> VCFHeteroHomoRecord:
	# 片親が不明の時、Genotypeを補う
	total_int_gt = 1 if pair == ParentComb.P00x01 else 3
	if not family.mat_known:
		pat_int_gt = Genotype.unphased(geno[1])
		mat_int_gt = total_int_gt - pat_int_gt
		if pat_int_gt == 3 or mat_int_gt < 0 or 2 < mat_int_gt:
			wrong_type = 'Modifiable'
		else:
			geno[0] = mat_int_gt
	elif not family.pat_known:
		mat_int_gt = Genotype.unphased(geno[0])
		pat_int_gt = total_int_gt - mat_int_gt
		if mat_int_gt == 3 or pat_int_gt < 0 or 2 < pat_int_gt:
			wrong_type = 'Modifiable'
		else:
			geno[1] = pat_int_gt
	return VCFHeteroHomoRecord(pos, geno, i, wrong_type, pair)

def classify_record(i: int, vcf: VCFFamily, td: TypeDeterminer,
					family: KnownFamily,
					heho_records: list[Optional[VCFHeteroHomoRecord]],
					other_records: list[Optional[VCFImpFamilyRecord]]) -> None:
	record = vcf.records[i]
	pos = record.pos
	geno = vcf.records[i].geno
	samples = vcf.samples
	wrong_type, pair = CR.classify_record(record, td, family.is_one_unknown())
	if pair.is_homohomo():
		record_ = VCFHomoHomoRecord(pos, geno, i, wrong_type, pair)
		other_records[i] = record_.impute()
	elif pair.is_heterohomo():
		heho_records[i] = create_heterohomo_record(pos, geno, family, i,
														wrong_type, pair)
	elif pair == ParentComb.P01x01:
		other_records[i] = VCFHeteroHeteroLiteRecord(pos, geno, i, wrong_type)
	else:		# 候補が無い
		other_records[i] = VCFJunkRecord(pos, geno, i, wrong_type)

def classify_records_parallel(v: tuple[VCFFamily, TypeDeterminer, KnownFamily,
									   list[Optional[VCFHeteroHomoRecord]],
									   list[Optional[VCFImpFamilyRecord]],
									   int, int]) -> None:
	vcf, td, family, heho_records, other_records, i0, num_threads = v
	for i in range(i0, len(vcf.records), num_threads):
		classify_record(i, vcf, td, family, heho_records, other_records)

# HeteroHomoだけ別にする
# このあとHeteroHomoだけ補完するから
# その他はVCFFillableにした後補完する
def classify_records(vcf: VCFFamily, family: KnownFamily,
						option: Option) -> tuple[list[VCFHeteroHomoRecord],
												 list[VCFImpFamilyRecord]]:
	td = CR.get_typedeterminer(vcf.num_progenies(), option.ratio)
	records1: list[Optional[VCFHeteroHomoRecord]] = [None] * len(vcf)
	records2: list[Optional[VCFImpFamilyRecord]] = [None] * len(vcf)
	args = [ (vcf, td, family, records1, records2, i, option.num_threads)
										for i in range(option.num_threads) ]
	for arg in args:
		classify_records_parallel(arg)
	
	heho_records: list[VCFHeteroHomoRecord] = []
	other_records: list[VCFImpFamilyRecord] = []
	for i in range(len(vcf)):
		r1 = records1[i]
		r2 = records2[i]
		if r1 is not None:
			heho_records.append(r1)
		elif r2 is not None:
			other_records.append(r2)
	
	return (heho_records, other_records)

def create_family_vcf(v: tuple[VCFSmall, Family]) -> VCFFamily:
	return VCFFamily.create(v[0], v[1].samples())

def create_family_vcfs(orig_vcf: VCFSmall, large_families: list[KnownFamily],
										num_threads: int) -> list[VCFFamily]:
	v = [ (orig_vcf, family) for family in large_families ]
	if num_threads > 1:
		with Pool(num_threads) as pool:
			family_vcfs = pool.map(create_family_vcf, v)
	else:
		family_vcfs = list(map(create_family_vcf, v))
	return family_vcfs

def collect_same_parent_families(families: list[KnownFamily]
								) -> list[tuple[str, list[tuple[int, int]]]]:
	dic = defaultdict(list)		# { parent: [(family index, parent index)] }
	for i, family in enumerate(families):
		dic[family.mat].append((i, 0))
		dic[family.pat].append((i, 1))
	return sorted((p, v) for p, v in dic.items() if len(v) > 1)

def sort_records(rs: list[tuple[list[VCFHeteroHomoRecord],
								list[VCFImpFamilyRecord], int]]
							) -> list[list[tuple[VCFImpFamilyRecord, int]]]:
	indices1 = [ rs1[-1].index if rs1 else 0 for rs1, _, __ in rs ]
	indices2 = [ rs2[-1].index if rs2 else 0 for _, rs2, __ in rs ]
	max_index = max(indices1 + indices2)
	recordss: list[list[tuple[VCFImpFamilyRecord, int]]] = \
								[ [] for _ in range(max_index + 1) ]
	for rs1, rs2, parent_index in rs:
		for r1 in rs1:
			if r1.parents_wrong_type == 'Right':
				recordss[r1.index].append((r1, parent_index))
		for r2 in rs2:
			if r2.pair.is_homohomo():
				recordss[r2.index].append((r2, parent_index))
	return recordss

def modify_00x11_each(rs: list[tuple[list[VCFHeteroHomoRecord],
									 list[VCFImpFamilyRecord], int]]) -> None:
	if len(rs[0][0]) == 0 and len(rs[0][1]) == 0:
		return
	recordss = sort_records(rs)
	for records in recordss:
		if len(records) < 2:
			continue
		VCFImpFamilyRecord.modify_00x11(records)

# 0/0 x 1/1のrecordで別の家系の親になっているとき、
# 他のホモ×ホモやヘテロ×ホモとGenotypeが違うとき、修正する
def modify_00x11(heho_recordss: list[list[VCFHeteroHomoRecord]],
				 other_recordss: list[list[VCFImpFamilyRecord]],
				 families: list[KnownFamily]) -> None:
	# { parent: [(family index, parent index)] }
	fams = collect_same_parent_families(families)
	# やり方を変える
	# どの家系のどちらの親かだけ持っておけばよい
	for _, v in fams:
		rs = [ (heho_recordss[fam_index], other_recordss[fam_index], p_index)
												for fam_index, p_index in v ]
		modify_00x11_each(rs)

# divide into hetero x homo and other types
def divide_vcf_into_record_types(family_vcfs: list[VCFFamily],
									families: list[KnownFamily], option: Option
									) -> tuple[list[list[VCFHeteroHomoRecord]],
											   list[list[VCFImpFamilyRecord]]]:
	heho_recordss: list[list[VCFHeteroHomoRecord]] = []
	other_recordss: list[list[VCFImpFamilyRecord]] = []
	for vcf, family in zip(family_vcfs, families):
		heho_records, other_records = classify_records(vcf, family, option)
		heho_recordss.append(heho_records)
		other_recordss.append(other_records)
	return (heho_recordss, other_recordss)

def fill_vcf(dic_vcfs: dict[str, list[VCFHeteroHomo]],
						other_recordss: list[list[VCFImpFamilyRecord]],
						families: list[KnownFamily]) -> list[VCFFillable]:
	# 同じ家系でまとめる
	family_indices = { family.parents(): i
							for i, family in enumerate(families) }
	vcfss_heho: list[list[VCFHeteroHomo]] = [ [] for _ in families ]
	for vcfs in dic_vcfs.values():
		for vcf in vcfs:
			index = family_indices[vcf.parents()]
			vcfss_heho[index].append(vcf)
	
	# 家系ごとに残ったレコードをcorrectする
	return [ VCFFillable.fill(vcfs, other_records)
				for vcfs, other_records in zip(vcfss_heho, other_recordss) ]

def impute_hetero_homo(orig_vcf: VCFSmall, families: list[KnownFamily],
									geno_map: Map, op: Option
									) -> tuple[list[list[VCFHeteroHomo]],
											   list[list[VCFImpFamilyRecord]]]:
	# create a VCF for each large family
	family_vcfs = create_family_vcfs(orig_vcf, families, op.num_threads)
	# classify records
	heho_recordss, other_recordss = divide_vcf_into_record_types(family_vcfs,
																families, op)
	# 家系間で調整するから家系をまとめて処理しなければならない
	modify_00x11(heho_recordss, other_recordss, families)
	
	# 家系ごとに処理する
	vcfss: list[list[VCFHeteroHomo]] = []
	for records, others, family in zip(heho_recordss, other_recordss, families):
		vcfs, unused = VCFHeteroHomo.impute_vcfs(records, family.samples(),
															geno_map, orig_vcf)
		vcfss.append(vcfs)
		others.extend(unused)
	
	return (vcfss, other_recordss)

def impute(orig_vcf: VCFSmall, families: list[KnownFamily],
							  geno_map: Map, option: Option) -> VCFGeno:
	vcfss, other_recordss = impute_hetero_homo(orig_vcf, families,
															geno_map, option)
	
	# 同じ親でVCFをまとめる
	dic_vcfs: dict[str, list[VCFHeteroHomo]] = defaultdict(list)
	for vcfs, family in zip(vcfss, families):
		for vcf in vcfs:
			if len(vcf) == 0:
				# 家系で一つもVCFが無いと家系がわからなるので、
				# 空でも追加しておく
				dic_vcfs[family.mat].append(vcf)
				dic_vcfs[family.pat].append(vcf)
			elif vcf.is_mat_hetero():
				dic_vcfs[family.mat].append(vcf)
			else:
				dic_vcfs[family.pat].append(vcf)
	
	v = sorted(dic_vcfs.items())
	for parent, vcfs in v:
		VCFHeteroHomo.inverse_phases(vcfs)
	
	filled_vcfs = fill_vcf(dic_vcfs, other_recordss, families)
	print("%d large families have been imputed." % len(families))
	return VCFFillable.merge(filled_vcfs, orig_vcf.get_samples())

__all__ = ['impute_large_normal_families']
